File: leetcode-go/Q042_Trapping_Rain_Water/Q042_Trapping_Rain_Water.go
package Q042_Trapping_Rain_Water

func trap(height []int) int {
	l, r := 0, len(height)-1
	lmax, rmax := 0, 0
	ans := 0
	for l < r {
		lmax = max(lmax, height[l])
		rmax = max(rmax, height[r])
		if lmax < rmax {
			ans += lmax-height[l]
			l++
		} else {
			ans += rmax-height[r]
			r--
		}
	}
	return ans
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

File: src/Y2018/facebook/Q042_Trapping_Rain_Water.java
package Y2018.facebook;

public class Q042_Trapping_Rain_Water {

    public int trap(int[] height) {
        int l = 0, r = height.length - 1;
        int lmax = height[l], rmax = height[r];
        int result = 0;
        while (l <= r) {
            if (height[l] < height[r]) {
                if (height[l] > lmax) lmax = height[l];
                else result += lmax - height[l];
                l++;
            } else {
                if (height[r] > rmax) rmax = height[r];
                else result += rmax - height[r];
                r--;
            }
        }
        return result;
    }

}
